src/js/react/useSyncExternalStoreWithSelector.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import {
+  useRef,
+  useEffect,
+  useMemo,
+  useDebugValue,
+  useSyncExternalStore,
```

Comment:
This is the reason we have a hard dependency on React 18 for this implementation. Targeting backwards compatibility before React 18 will require additional, non-trivial complexity.